cpp/ql/test/library-tests/ir/ssa/aliased_ssa_overlap.expected

```diff
@@ -0,0 +1,542 @@
+| ssa.cpp:322:8:322:11 | ?ptr1[0..?)<unknown> | ssa.cpp:322:8:322:11 | ptr1 | MayPartiallyOverlap |
+| ssa.cpp:322:8:322:11 | ?ptr1[0..?)<unknown> | ssa.cpp:322:8:322:11 | ptr1[0..?)<unknown> | MayPartiallyOverlap |
+| ssa.cpp:322:8:322:11 | ptr1 | ssa.cpp:322:8:322:11 | ?ptr1[0..?)<unknown> | MustTotallyOverlap |
+| ssa.cpp:322:8:322:11 | ptr1 | ssa.cpp:322:8:322:11 | ptr1 | MustExactlyOverlap |
+| ssa.cpp:322:8:322:11 | ptr1 | ssa.cpp:322:8:322:11 | ptr1[0..?)<unknown> | MustTotallyOverlap |
+| ssa.cpp:322:8:322:11 | ptr1[0..?)<unknown> | ssa.cpp:322:8:322:11 | ?ptr1[0..?)<unknown> | MayPartiallyOverlap |
+| ssa.cpp:322:8:322:11 | ptr1[0..?)<unknown> | ssa.cpp:322:8:322:11 | ptr1 | MayPartiallyOverlap |
+| ssa.cpp:322:8:322:11 | ptr1[0..?)<unknown> | ssa.cpp:322:8:322:11 | ptr1[0..?)<unknown> | MayPartiallyOverlap |
+| ssa.cpp:402:7:402:7 | b | ssa.cpp:402:7:402:7 | b | MustExactlyOverlap |
+| ssa.cpp:402:7:402:7 | b | ssa.cpp:402:7:402:7 | b[0..4)<int> | MustTotallyOverlap |
+| ssa.cpp:402:7:402:7 | b[0..4)<int> | ssa.cpp:402:7:402:7 | b | MayPartiallyOverlap |
+| ssa.cpp:402:7:402:7 | b[0..4)<int> | ssa.cpp:402:7:402:7 | b[0..4)<int> | MustExactlyOverlap |
+| ssa.cpp:403:7:403:7 | c | ssa.cpp:403:7:403:7 | c | MustExactlyOverlap |
+| ssa.cpp:418:7:418:7 | x | ssa.cpp:418:7:418:7 | x | MustExactlyOverlap |
+| ssa.cpp:418:7:418:7 | x | ssa.cpp:418:7:418:7 | x[0..4)<int> | MustTotallyOverlap |
+| ssa.cpp:418:7:418:7 | x | ssa.cpp:418:7:418:7 | x[4..8)<int> | MustTotallyOverlap |
+| ssa.cpp:418:7:418:7 | x[0..4)<int> | ssa.cpp:418:7:418:7 | x | MayPartiallyOverlap |
+| ssa.cpp:418:7:418:7 | x[0..4)<int> | ssa.cpp:418:7:418:7 | x[0..4)<int> | MustExactlyOverlap |
+| ssa.cpp:418:7:418:7 | x[4..8)<int> | ssa.cpp:418:7:418:7 | x | MayPartiallyOverlap |
+| ssa.cpp:418:7:418:7 | x[4..8)<int> | ssa.cpp:418:7:418:7 | x[4..8)<int> | MustExactlyOverlap |
+| ssa.cpp:442:37:442:37 | *a | ssa.cpp:442:37:442:37 | *a | MustExactlyOverlap |
+| ssa.cpp:442:37:442:37 | *a | ssa.cpp:442:37:442:37 | ?*a | MustTotallyOverlap |
+| ssa.cpp:442:37:442:37 | ?*a | ssa.cpp:442:37:442:37 | *a | MayPartiallyOverlap |
+| ssa.cpp:442:37:442:37 | ?*a | ssa.cpp:442:37:442:37 | ?*a | MayPartiallyOverlap |
+| ssa.cpp:442:46:442:46 | *b | ssa.cpp:442:46:442:46 | *b | MustExactlyOverlap |
+| ssa.cpp:442:46:442:46 | *b | ssa.cpp:442:46:442:46 | ?*b | MustTotallyOverlap |
+| ssa.cpp:442:46:442:46 | ?*b | ssa.cpp:442:46:442:46 | *b | MayPartiallyOverlap |
+| ssa.cpp:442:46:442:46 | ?*b | ssa.cpp:442:46:442:46 | ?*b | MayPartiallyOverlap |
+| ssa.cpp:446:34:446:34 | *a | ssa.cpp:446:34:446:34 | *a | MustExactlyOverlap |
+| ssa.cpp:446:34:446:34 | *a | ssa.cpp:446:34:446:34 | *a[0..?)<unknown> | MustTotallyOverlap |
+| ssa.cpp:446:34:446:34 | *a | ssa.cpp:446:34:446:34 | ?*a | MustTotallyOverlap |
+| ssa.cpp:446:34:446:34 | *a | ssa.cpp:446:34:446:34 | ?*a[0..?)<unknown> | MustTotallyOverlap |
+| ssa.cpp:446:34:446:34 | *a | ssa.cpp:446:43:446:43 | *b | MustExactlyOverlap |
```

Comment:
I am confused by this. Why are `*a` and `*b` exactly overlapping? That means that a write `*a` will be fully read by a read of `*b`, but looking at the test you added that is clearly not the case. So I think something is wrong here. Although, I dont think your fix is the right one. I will look a bit into this today!

Reply:
Note that these results come from the commit with my fix removed (i.e. as `main`), so it's an existing problem.  I would argue that my change improves things by removing these rows, but I *wouldn't* argue that removing them is necessarily the best fix possible.

Reply:
Yeah, I agree this is not the fault of your changes. It has been around for a very long time. I managed to dig up an old comment from Jonas on one of my first PRs here: https://github.com/github/codeql/pull/2772#discussion_r375428192 😄

I do agree that your changes makes this slightly better in this specific case, indeed

Reply:
If it turns out that escaped `EntireAllocationMemoryLocation` always _exactly_ overlap as specified [here](https://github.com/github/codeql/blob/d1cca13563ff9396f01e1bb6670d4c35e78cef58/cpp/ql/lib/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll#L686-L688) then I think I don't fully understand `EntireAllocationMemoryLocation`. As written in https://github.com/github/codeql/pull/2696 it's supposed to represent "access to the entire contiguous allocation, without knowing exactly how big that allocation is", and I don't see why two such accesses should exactly overlap. It may be worth discussing this with Dave since he was the one that introduced these memory locations

Reply:
IIRC:

An allocation is a contiguous chunk of memory, disjoint from all other allocations. We create an allocation for each local variable, each parameter, and (in aliased SSA) the memory pointed to by each pointer or reference parameter. We might create an allocation for each call to `malloc`/`operator new`/etc., but I can't remember. Sometimes we know the exact size of an allocation (`int x` creates a 4-byte allocation for `x`); sometimes we don't (`void foo(int* p)` creates an allocation for `*p` (the memory pointed to by `p`), but that allocation has unknown size.

A location a contiguous chunk of memory accessed by a particular instruction, either directly (load, store) or indirectly (side effect of a call). A location lives in an allocation. Sometimes we know the exact offset and size of that location inside of the allocation (in `int x; x = 5;`, the value is stored to the 4-byte location at offset 0 from the start of the `x` allocation, which is of course the whole allocation). Sometimes we just know that the location is somewhere in a specific allocation, but not exactly where (in `char a[10]; a[i] = 5;`, the value is stored to a 1-byte location in the allocator for `a`, but at an unknown offset).

An `EntireAllocationMemoryLocation` is a location that represents an entire specific allocation, without having to know the exact size of the allocation. We use it when "initializing" indirect parameters at the start of the function (`void foo(int* p)` will initialize the memory pointed to by `p` with an `InitializeIndirection` instruction, using an `EntireAllocationMemoryLocation`, representing whatever the caller did to populate that memory). We also use it at the end of the function on the `ReturnIndirection` instruction, to represent that the caller could eventually read all of that indirect memory.

We assume the size of an allocation doesn't change, so we know that two accesses to the `EntireAllocationMemoryLocation` for the same allocation access the same range of memory. We don't know what that size is, but we know that the sizes of both accesses are the same.

`MustExactlyOverlap` is one of the four possible overlap releationships between two memory accesses `a` and `b`:
- `MustExactlyOverlap` - `a` and `b` access exactly the same set of bits (same location, same size). We also require that `a` and `b` have the same type. This is the common case with scalar variables: In `int x = 5; return x;` the store to `x` and the load from `x` both access the same four bytes of `x`. This lets us link the use (load) to the definition (store) in SSA as an "exact definition". Similarly, in `int x = 5; x = 7; return x;`, we know that the second store to `x` exactly overlaps the first store to `x`, so the first store cannot reach any use after the second store.
- `MustTotallyOverlap` - `a` accesses the same set of bits as `b`, plus maybe more. This most often happens with structs. For something like `Point p = { 1, 2 }; return p.x;`, the initialization of `p` `MustTotallyOverlap`s the use of the field of `p`, because the initialization writes to all the bits of `p.x` plus another field of `p`. This lets us link use of `p.x` to the definition of `p`, but it's not an _exact_ definition because the accesses are different sizes and types. Similarly, in `p.x = 5; p = q;`, the second store `MustTotallyOverlap`s the first store, so know that nothing after the second store can see the first store.
- `MayPartiallyOverlap` - `a` and `b` might have at least one bit in common, but they might not totally or exactly overlap. Or they might not overlap at all. The common case of this is when there might be aliasing, but we're not sure. In `*p = 5; *q = 7; return *p;`, if we don't have any special knowledge of what `p` and `q` point to, we have to assume that they might point to the same location, or they might point to disjoint locations, or, in weird cases, `p` might point halfway into the memory that `q` points to. When we encounter this kind of relationship when building SSA, we have to model that the load of `*p` in `return *p` might get its value from the store to `*q` (if `p` and `q` point to the same address), might get its value from the store to `*p` (if `p` and `q` point to disjoint locations), or some combination of both (if `p` and `q` point to different addresses within less than 4 bytes of one another). This is where aliased SSA gets weird and `Chi` nodes get involved. That's a whole separate topic, though.
- No overlap - We know that `a` and `b` are entirely disjoint: In `int x = 5; int y = 6;`, the two stores are known to not overlap, because they access two different allocations.

Reply:
> We assume the size of an allocation doesn't change, so we know that two accesses to the EntireAllocationMemoryLocation for the same allocation access the same range of memory.

But on main we don't enforce that they have the same allocation. We only enforce that they have the same virtual variable (see the disjunct linked above: https://github.com/github/codeql/blob/d1cca13563ff9396f01e1bb6670d4c35e78cef58/cpp/ql/lib/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll). Requiring that they have the same allocation is exactly the proposed change that Geoffrey is making in this PR.

What I would like to ubderstsnf is whether that's a bug that's always been there which we are now fixing, or whether there is something about EntireAllocationMemoryLocation that I don't understand that justifies only comparing virtual vatiables

Reply:
Thank you for the explanation @dbartol, that's cleared things up a lot for me.

So an `Allocation` is the top of the hierarchy, a `MemoryLocation` is an area inside one, and a `VirtualVariable` is a `MemoryLocation` representing a set of locations "under" it that may overlap.  Thus a `VirtualVariable` is itself within one `Allocation`.  Thus `def.getVirtualVariable() = use.getVirtualVariable()` *should* imply that `def.getAllocation() = use.getAllocation()` ... but apparently in some cases it doesn't?

Reply:
> Thus `def.getVirtualVariable() = use.getVirtualVariable()` _should_ imply that `def.getAllocation() = use.getAllocation()` ... but apparently in some cases it doesn't?

Not when an an allocation escapes. Because when an allocation escapes the result of `getVirtualVariable()` is `AllAliasedMemory` (and there is only one `AllAliasedMemory` per function).

So when `def` and `use` dont escape `def.getVirtualVariable() = use.getVirtualVariable()` indeed implies that they have the same allocation (since an `EntireAllocationMemoryLocation` is its own virtual variable in that case), but when two `EntireAllocationMemoryLocation` in the same enclosing function escape they will exactly overlap according as per https://github.com/github/codeql/blob/d1cca13563ff9396f01e1bb6670d4c35e78cef58/cpp/ql/lib/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll#L686-L688.

Reply:
A virtual variable is a set of memory locations where no location within that virtual variable can overlap with any location outside that virtual variable. This lets us compute the overlap _within_ each virtual variable independently.

Suppose all of your memory locations were scalar local variables, and there was no aliasing. In that case, when building SSA, you would find the definition for a use by looking back for the most recent definition of the same variable, ignoring definitions of all other variables because they can't possibly overlap with the use. Virtual variables let us do something similar even in the presence of aliasing, by grouping all memory locations into disjoint sets. When looking for the definition of a particular use, we look for the most recent definition that writes to the same virtual variable as the use, ignoring any definitions that write to other virtual variables.

IIRC, we create one virtual variable for each unaliased local variable and parameter, plus a single "All aliased memory" virtual variable that includes everything else.

Reply:
So based on all the above the piece I'm missing, I think, is that `AllAliasedMemory` can in fact be associated with more than one `Allocation`? Perhaps any `VirtualVariable` can be?  I'm having a hard time reconciling this with the original idea that "A location lives in an allocation" and that `VirtualVariable extends MemoryLocation`, so a `VirtualVariable` *ought* to live in an allocation as well.

Reply:
Dave didn't get all the details, no. Let me write out some more stuff:

`AllAliasedMemory` isn't associated with any allocation. An `AllAliasedMemory` doesn't know about allocations. The definition is [here](https://github.com/github/codeql/blob/4c8aec0db529870c2d4b1f6158a76c10348437ed/cpp/ql/lib/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll#L227C4-L227C9). Note that, unlike [VariableAllocation](https://github.com/github/codeql/blob/4c8aec0db529870c2d4b1f6158a76c10348437ed/cpp/ql/lib/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll#L202) (which is the one Dave was talking about), `AllAliasedMemory` it's not defined in terms of `Allocation`.

So in general a location does not "live inside an allocation". Some `MemoryLocation`s are defined in terms of a specific allocation (`VariableMemoryLocation` and `EntireAllocationMemoryLocation` are defined in terms of an allocation, and `GroupedMemoryLocation` are defined in terms of a set of allocations).

Reply:
@dbartol this question/comment may have disappeared in the conversation: https://github.com/github/codeql/pull/17225#discussion_r1751045351. I'd really like to hear your answer to this one 🥺

Reply:
Oh, sorry, missed that. That's a bug.

The obvious fix would be to check that the allocation also matches before saying that the two `EntireAllocationMemoryLocation`s overlap. That should be safe, but I think there's an underlying problem: I believe the current code assumes that no two `EntireAllocationMemoryLocation`s share the same virtual variable. By design, we unsoundly assume that the memory pointed to by two different pointer parameters doesn't overlap, so we create a separate virtual variable for each indirect parameter.

In the suspicious test case, are the two memory accesses accessing the same indirect parameter? Or are we conflating the `EntireAllocationMemoryLocation`s from two different indirect parameters? If it's the latter, I think that means we've managed to assign at least one of the `EntireMemoryAllocation` objects to multiple virtual variables.

Reply:
I still need some more help on this, I've been thrown in at the deep end here.  I think @dbartol is saying that my change is correct, but that there might be other places where a similar change is required to distinguish `EntireAllocationMemoryLocation`s within the same virtual variable?  Do you have any idea where?  I guess I can search all uses of `EntireAllocationMemoryLocation` if this is what we're looking for?

Reply:
Yeah, Dave is saying that the change should be safe, but that the change shouldn't be necessary to do here since the current code assumes that two `EntireAllocationMemoryLocations` never have the same virtual variable.

I'm not sure that's true, though: An `EntireAllocationMemoryLocation` is an `AllocationMemoryLocation` (see [here](https://github.com/github/codeql/blob/9250c29bd566f839b846c1c72c1c41063654ad3d/cpp/ql/lib/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll#L485)), and the virtual variable of an `AllocationMemoryLocation` is `AllAliasedMemory` when the allocation escapes (see [here](https://github.com/github/codeql/blob/9250c29bd566f839b846c1c72c1c41063654ad3d/cpp/ql/lib/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll#L307)). So the assumption that two `EntireAllocationMemoryLocation` never share the same virtual variable is definitely false.

So either we need:
- @dbartol to explain to use which code makes this assumption (and fix that code), or
- to change the definition of `getVirtualVariable` for `EntireAllocationMemoryLocation` so that it always returns `this` even if the allocation escapes.

The latter of those two will then result in `def.getVirtualVariable() = use.getVirtualVariable()` implying that they have the same allocation (which means the change you are suggesting here will be enforced automatically).

Reply:
Cc @dbartol on ^

Reply:
I'm no longer convinced that we assume that two different `EntireAllocationMemoryLocation`s are never part of the same virtual variable. I believe that we try to treat the extra "variable" we introduce for pointer parameters like any other _local_ variable: If its address never escapes, then it gets its own virtual variable. If its address does escape, it gets grouped into the "all aliased memory" virtual variable.

So, two different `EntireAllocationMemoryLocation`s should never overlap, even if they are in the same virtual variable. I think this just means changing:

```ql
use instanceof EntireMemoryAllocation and
result instanceof MustExactlyOverlap
```
to
```ql
use = def and
result instanceof MustExactlyOverlap
```

I'm having trouble coming up with a test case that would show the difference in behavior, though.